engine/util/rng.hpp

```diff
@@ -194,66 +420,82 @@ double basic_rng_t<Engine>::range( double min, double max )
  *     copyright notice is preserved.
  */
 template <typename Engine>
-double basic_rng_t<Engine>::gauss( double mean, double stddev, bool truncate_low_end )
+double basic_rng_t<Engine>::gauss( double mean, double stddev )
 {
   assert(stddev >= 0 && "Calling gauss with negative stddev");
 
-  double z;
+  if ( stddev == 0 )
+    return mean;
 
-  if ( stddev != 0 )
+  if ( gauss_pair_use )
   {
-    if ( gauss_pair_use )
-    {
-      z = gauss_pair_value;
-      gauss_pair_use = false;
-    }
-    else
-    {
-      double x1, x2, w, y1, y2;
-      do
-      {
-        x1 = 2.0 * real() - 1.0;
-        x2 = 2.0 * real() - 1.0;
-        w = x1 * x1 + x2 * x2;
-      }
-      while ( w >= 1.0 || w == 0.0 );
-
-      w = std::sqrt( ( -2.0 * std::log( w ) ) / w );
-      y1 = x1 * w;
-      y2 = x2 * w;
-
-      z = y1;
-      gauss_pair_value = y2;
-      gauss_pair_use = true;
-    }
+    gauss_pair_use = false;
+    return mean + gauss_pair_value * stddev;
   }
-  else
-    z = 0.0;
 
-  double result = mean + z * stddev;
+  double x1, x2, w, y1, y2;
+  do
+  {
+    x1 = 2.0 * real() - 1.0;
+    x2 = 2.0 * real() - 1.0;
+    w = x1 * x1 + x2 * x2;
+  }
+  while ( w >= 1.0 || w == 0.0 );
+
+  w = std::sqrt( ( -2.0 * std::log( w ) ) / w );
+  y1 = x1 * w;
+  y2 = x2 * w;
 
-  // True gaussian distribution can of course yield any number at some probability.  So truncate on the low end.
-  if ( truncate_low_end && result < 0 )
-    result = 0;
+  gauss_pair_value = y2;
+  gauss_pair_use = true;
 
-  return result;
+  return mean + y1 * stddev;
+}
+
+template <typename Engine>
+double basic_rng_t<Engine>::gauss_ab( double mean, double stddev, double min, double max )
+{
+  assert( stddev >= 0.0 && "Stddev must be non-negative." );
+  assert( min < max && "Minimum must be less than maximum." );
+  assert( mean >= min && mean <= max && "Mean must be contained within the interval [min, max]" );
+
+  if ( stddev == 0.0 )
+    return mean;
+
+  double min_cdf      = stdnormal_cdf( ( min - mean ) / stddev );
+  double max_cdf      = stdnormal_cdf( ( max - mean ) / stddev );
```

Comment:
I wonder how much time could be saved in some cases by not calculating these CDF values every time the function is called for the cases where `min` and `max` are going to be constant for the entire simulation.

Given that (a) I think it would require the caller to do things very differently and (b) I don't think the rate that this is called is going to be that high compared to things like `real`, it's probably not worth doing at the moment, but I thought it was worth pointing out.

Reply:
I agree that caching `stdnormal_cdf` values could potentially be valuable, and further profiling could be done on that front w.r.t the expense of `stdnormal_cdf`.